### PR TITLE
ecs-app supports network app_type (and multiple TCP and TCP(s) ports)

### PR DESCRIPTION
**Pull request by igorkotof**
image
image
image
image

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -47,7 +41,7 @@ module "alb" {
         timeout             = var.alb_health_check_timeout
         matcher             = var.alb_health_check_valid_response_codes
         port                = "traffic-port"
-        protocol            = "HTTP"
+        protocol            = var.app_type == "web" ? "HTTP" : "TCP"
```
Comment (DimmKirr) [inline main.tf]: @igorkotof (can't comment there)
How would `path                = var.alb_health_check_path` work? Should we use locals to set a path specific for TCP?

File: main.tf [hunk 2]
```diff
@@ -1,39 +1,33 @@
...
-      }
-    ]) : []
+  [
+    {
+      port               = 443
+      protocol           = var.app_type == "web" ? "HTTPS" : "TLS"
+      certificate_arn    = var.tls_cert_arn
+      target_group_index = 0
+    }
+  ]) : []
```
Comment (DimmKirr) [inline main.tf:26]: @igorkotof How would it work with multiple ports in a target group? (like 4444,4445 etc)?

Comment (igorkotof) [reply main.tf:26]: @AutomationD pls take a look into: `http_tcp_listeners = local.http_tcp_listeners`

File: locals.tf [hunk 3]
```diff
@@ -102,4 +102,86 @@ locals {
     ] : [],
     (var.datadog_enabled && var.ecs_launch_type == "EC2") ? module.datadog.volumes : []
   )
+
+  http_tcp_listeners = var.app_type == "tcp-app" ? [
+    for index, portmap in var.port_mappings :
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
    for index, port_mapping in var.port_mappings :
```

File: locals.tf [hunk 4]
```diff
@@ -102,4 +102,86 @@ locals {
...
+      {
+        port               = portmap.host_port
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
        port               = port_mapping.host_port
```

File: locals.tf [hunk 5]
```diff
@@ -102,4 +102,86 @@ locals {
...
+      }
+    ] : [
+      {
+        port               = var.http_port
+        protocol           = "HTTP"
+        target_group_index = 0
+      },]
+
+  ecs_service_tcp_port_mappings = [
+    for index, portmap in var.port_mappings :
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
    for index, port_mapping in var.port_mappings :
```

File: locals.tf [hunk 6]
```diff
@@ -102,4 +102,86 @@ locals {
...
+      {
+        container_name      = var.name
+        container_port      = portmap.container_port
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
        container_port      = port_mapping.container_port
```

File: locals.tf [hunk 7]
```diff
@@ -102,4 +102,86 @@ locals {
...
+        host_port           = portmap.host_port
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
        host_port           = port_mapping.host_port
```

File: locals.tf [hunk 8]
```diff
@@ -102,4 +102,86 @@ locals {
...
+      tags = {
+        Name = var.name
+        env  = var.env
+        app  = local.name
+      }
+    }
+  ]
+
+  target_groups_tcp = [
+    for portmap in var.port_mappings :
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
    for port_mapping in var.port_mappings :
```

File: locals.tf [hunk 9]
```diff
@@ -102,4 +102,86 @@ locals {
...
+    {
+      name_prefix          = local.name_prefix
+      backend_protocol     = "TCP"
+      backend_port         = portmap.container_port
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
      backend_port         = port_mapping.container_port
```

File: locals.tf [hunk 10]
```diff
@@ -102,4 +102,86 @@ locals {
...
+
+      health_check = {
+        enabled             = true
+        interval            = var.alb_health_check_interval
+        path                = null
+        healthy_threshold   = var.alb_health_check_healthy_threshold
+        unhealthy_threshold = var.alb_health_check_unhealthy_threshold
+        timeout             = null
+        matcher             = null
+        port                = portmap.host_port
```
Comment (DimmKirr) [inline locals.tf]:
```suggestion
        port                = port_mapping.host_port
```

File: ecs-modules/ecs-service/variables.tf [hunk 11]
```diff
@@ -14,8 +14,8 @@ variable "app_type" {
   default     = "web"
 
   validation {
-    condition     = var.app_type == "web" || var.app_type == "worker"
-    error_message = "The app_type value must be a valid type: web or worker."
+    condition     = var.app_type == "web" || var.app_type == "tcp-app" || var.app_type == "worker"
+    error_message = "The app_type value must be a valid type: web or worker or tcp-app."
```
Comment (DimmKirr) [inline ecs-modules/ecs-service/variables.tf]:
```suggestion
    error_message = "The app_type value must be a valid type: web, worker or tcp-app."
```

File: ecs-modules/ecs-service/main.tf [hunk 12]
```diff
@@ -105,7 +105,7 @@ resource "aws_ecs_service" "this" {
   }
 
   dynamic "load_balancer" {
-    for_each = var.app_type == "web" ? [
+    for_each = var.app_type == "web" || var.app_type == "tcp-app" ? [
     for p in var.port_mappings : {
```
Comment (DimmKirr) [inline ecs-modules/ecs-service/main.tf]:
```suggestion
    for port_mapping in var.port_mappings : {
```

File: ecs-modules/ecs-service/main.tf [hunk 13]
```diff
@@ -105,7 +105,7 @@ resource "aws_ecs_service" "this" {
...
       container_name   = p.container_name
       target_group_arn = p.target_group_arn
```
Comment (DimmKirr) [inline ecs-modules/ecs-service/main.tf]:
```suggestion
      target_group_arn = port_mapping.target_group_arn
```

File: ecs-modules/ecs-service/main.tf [hunk 14]
```diff
@@ -105,7 +105,7 @@ resource "aws_ecs_service" "this" {
   }
 
   dynamic "load_balancer" {
-    for_each = var.app_type == "web" ? [
+    for_each = var.app_type == "web" || var.app_type == "tcp-app" ? [
     for p in var.port_mappings : {
       container_name   = p.container_name
```
Comment (DimmKirr) [inline ecs-modules/ecs-service/main.tf]:
```suggestion
      container_name   = port_mapping.container_name
```

File: locals.tf [hunk 15]
```diff
@@ -102,4 +102,86 @@ locals {
...
+    for index, portmap in var.port_mappings :
+      {
+        container_name      = var.name
+        container_port      = portmap.container_port
+        host_port           = portmap.host_port
+        target_group_arn    = length(module.alb[*].target_group_arns) >= 1 ? module.alb[0].target_group_arns[index] : ""
+      }
+    ]
+
+  target_groups_web = [
```
Comment (DimmKirr) [inline locals.tf:130]: Would this mean that target_groups_web will always be present? Do we need to  add a ternary here?

File: locals.tf [hunk 16]
```diff
@@ -102,4 +102,86 @@ locals {
...
+        matcher             = null
+        port                = portmap.host_port
+        protocol            = "TCP"
+      }
+
+      tags = {
+        Name = var.name
+        env  = var.env
+        app  = local.name
+      }
```
Comment (DimmKirr) [inline locals.tf]: Provided by global tags
```suggestion
```

File: main.tf [hunk 17]
```diff
@@ -57,6 +56,7 @@ module "alb" {
       }
     }
   ])
+>>>>>>> ad9062d8fae00429f7d9c96164fd275582638f2e
```
Comment (DimmKirr) [inline main.tf]: 🤔

File: variables.tf [hunk 18]
```diff
@@ -19,8 +19,8 @@ variable "app_type" {
   default     = "web"
 
   validation {
-    condition     = var.app_type == "web" || var.app_type == "worker"
-    error_message = "The app_type value must be a valid type: web or worker."
+    condition     = var.app_type == "web" || var.app_type == "tcp-app" || var.app_type == "worker"
+    error_message = "The app_type value must be a valid type: web or worker or tcp-app."
```
Comment (DimmKirr) [inline variables.tf]:
```suggestion
    error_message = "The app_type value must be a valid type: web, worker or tcp-app."
```

File: locals.tf [hunk 19]
```diff
@@ -102,4 +102,86 @@ locals {
...
+        port                = "traffic-port"
+        protocol            = "HTTP"
+      }
+
+      tags = {
+        Name = var.name
+        env  = var.env
+        app  = local.name
+      }
+    }
```
Comment (DimmKirr) [inline locals.tf]: Provided by global tags.
```suggestion
```

**Review by DimmKirr — CHANGES_REQUESTED** [review]
Looks good, let's address items. 

In addition it would be great to see tests with:
- Simple two-port tcp application on EC2 (selenium hub or nginx tcp proxy to github:22?)
- Simple two-port tcp application on Fargate (selenium hub or nginx tcp proxy to github:22?)

File: locals.tf (hunk 15 above)
Comment (igorkotof) [reply locals.tf:130]: @AutomationD it's just local.tf, condition is here https://github.com/hazelops/terraform-aws-ecs-app/blob/818bdeacefd3ee29ecd47c6ef2c63a3498756541/main.tf#L28

**Review by DimmKirr — APPROVED** [review]
Let's cleanup commits and update missing port_mappings.

Otherwise looks good!

**Pull request merged**